### ISSUE TITLE
ChromeDriver could not be found 

### ISSUE THREAD

**Issue by dcsan** [wechaty/wechaty#164]
This is just an information since others have the same problem... or if you know the reason.

I setup wechaty on two new machines (showing it to a friend) and both times hit this error on trying to run the demo. However I don't always hit this problem - other machines have been OK.

> ChromeDriver could not be found

```
INFO Wechaty v#git[044d24f 0.6.67] initializing...
ERR PuppetWebBrowserDriver initChromeDriver() Wechaty require `chrome-driver` to be installed.
ERR PuppetWebBrowser init() exception: The ChromeDriver could not be found on the current PATH. Please download the latest version of the ChromeDriver from http://chromedriver.storage.googleapis.com/index.html and ensure it can be found on your PATH.
WARN PuppetWebBrowser quit() exception: Cannot read property 'close' of undefined
WARN PuppetWebBrowser driver.quit() exception: Cannot read property 'close' of undefined
ERR PuppetWeb initBrowser() exception: The ChromeDriver could not be found on the current PATH. Please download the latest version of the ChromeDriver from http://chromedriver.storage.googleapis.com/index.html and ensure it can be found on your PATH.
ERR PuppetWeb init() exception: Error: The ChromeDriver could not be found on the current PATH. Please download the latest version of the ChromeDriver from http://chromedriver.storage.googleapis.com/index.html and ensure it can be found on your PATH.
    at Error (native)
    at ServiceBuilder (/Users/dc/dev/rikai/ta-bot/wechaty/node_modules/selenium-webdriver/chrome.js:191:13)
    at getDefaultService (/Users/dc/dev/rikai/ta-bot/wechaty/node_modules/selenium-webdriver/chrome.js:280:22)
    at Function.createSession (/Users/dc/dev/rikai/ta-bot/wechaty/node_modules/selenium-webdriver/chrome.js:700:44)
    at createDriver (/Users/dc/dev/rikai/ta-bot/wechaty/node_modules/selenium-webdriver/index.js:167:33)
    at Builder.build (/Users/dc/dev/rikai/ta-bot/wechaty/node_modules/selenium-webdriver/index.js:639:16)
    at BrowserDriver.<anonymous> (/Users/dc/dev/rikai/ta-bot/wechaty/src/puppet-web/browser-driver.js:79:26)
    at next (native)
    at /Users/dc/dev/rikai/ta-bot/wechaty/src/puppet-web/browser-driver.js:7:65
    at __awaiter (/Users/dc/dev/rikai/ta-bot/wechaty/src/puppet-web/browser-driver.js:3:12)
INFO Bot error: Error: The ChromeDriver could not be found on the current PATH. Please download the latest version of the ChromeDriver from http://chromedriver.storage.googleapis.com/index.html and ensure it can be found on your PATH.
(node:21414) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 2): Error: can not say before login
(node:21414) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 3): TypeError: Cannot read property 'quit' of undefined
```

To fix this i reinstalled just

> npm i chromedriver

and everything was OK.

Now I noticed the URL points to a google domain. 
> http://chromedriver.storage.googleapis.com

But since it came from (c)npm i was able to install OK.

**Comment by zixia** [comment]
This is expected because I remove the `chromedriver` dependence in wechaty at https://github.com/wechaty/wechaty/commit/9e7164f5a02ad7a7c952817c0af690b2645977c0

The reason I do this is because we will have many heads in the future, now we have two: phantomjs and chromedriver, later maybe we need android SDK/Emulator, etc. So I think to let the user install the exact dependence instead of we install everything is a better choice.

Also, there's a problem when you in China: you can use CDN from Taobao for npm, but chromedriver is hosted at another google server, so you have to set another proxy for installing chromedriver. It's will block some user to install Wechaty.

Finally, if you think we should include the chromedriver in Wechaty, we can discuss it as well. Because I'm not very sure if we should include it or not. Thanks.

**Comment by dcsan** [comment]
I think this needs to be documented then, as currently it will (sometimes?) leave an incomplete install. One way would be an installer/setup script, but that would need to also maybe support windows which would be a pain.

Is there a way to use PhantomJS in development, or is that just for testing? I tried:

```
const TestConfig = {
  "wechaty": {
    "profile": "TABOT",
    "DEFAULT_HEAD": "phantomjs", // chrome
    "DEFAULT_PUPPET": "web",
    "DEFAULT_PROFILE": "demo",
    "DEFAULT_PORT": 8788,
    "DEFAULT_PROTOCOL": "io|0.0.1",
    "DEFAULT_TOKEN": "WECHATY_IO_TOKEN",
    "DEFAULT_APIHOST": "api.wechaty.io",
    "CMD_CHROMIUM": "/wechaty/bin/xvfb-chromium"
  }
}

const WcBot = Wechaty.instance( TestConfig.wechaty );
```

**Comment by zixia** [comment]
Yes, I agree that this should be documented, because it will **always** leave an incomplete install.

Wechaty should be able to use PhantomJS in development, but as I'm not using it for months, there might have some issue, see also #60 #163

**Comment by eastmoon** [comment]
I have same problem.

Today, i need find some solution with WeChat bot.
And, i try wechaty by below step.

1 npm install --save wechaty
2 npm install chromedriver
3 create mybot.js and copy code : 6 lines of JavaScrip
4 node mybot.js

Then, i have same problem with wechaty could not find chromedriver.
For node.js, i am newbie, i use all i know to find answers, but still not idea how to solve.

**Comment by zixia** [comment]
Hi @eastmoon ,

### TL;DR

I suggest you install chromedriver globally by run `npm install -g chromedriver`, then you can find chromedriver in your PATH.

### Detail

Your issue is because `chromedriver` is not in your current PATH, so you can not run it.

After you ran `npm install chromedriver`, the chromedriver will be placed at `./node_modules/.bin/`, but this directory is not in your PATH by default.

As a result, Wechaty will also not be able to find it.

**Comment by dcsan** [comment]
This issue came up in another chatroom today too. It seems a frequent issue

**Comment by AaronZhang9511** [comment]
"npm install -g chrome" works on ubuntu 16.04
but as to centOS 7, it didn't work.

**Comment by zixia** [comment]
@dcsan Yes, I agree with you.

I'm planing to put the `chromedriver` detection in `npm run doctor`: if there's no chromedriver found in the system, the `npm run doctor` will print a user friendly error message, and the suggestion of how to fix that.

**Comment by zixia** [comment]
@dcsan @eastmoon at last,  I decided to restore chromedriver dependence, because it seems the simplest solution for this kind of issue. :P

**Comment by eastmoon** [comment]
@zixia Good decided.
I will try again in a few day late.
Hope this issue will be fix.

**Comment by brianwfl88** [comment]
Hi, I am having the same issue like eastmoon

I can run chromedriver command just fine, and also added chromedriver PATH from global scope.

**Comment by ax4** [comment]
Same as @brianwfl88 

I can run chromedriver in cmd, it's already in PATH. But it still fail at ChromeDriver.

**Comment by zixia** [comment]
Hi @brianwfl88 @ax4 ,

It seems you guys are both using windows.

Could you please provide the follow information for diagnosing?

1. the source code of your program
2. the command line command of starting your program
3. the full output of logs with `WECHATY_LOG=verbose`

**Comment by ax4** [comment]
I just copy from README in my `app.js`
```
const { Wechaty } = require('wechaty')

Wechaty.instance() // Singleton
    .on('scan', (url, code) => console.log(`Scan QR Code to login: ${code}\n${url}`))
    .on('login', user => console.log(`User ${user} logined`))
    .on('message', message => console.log(`Message: ${message}`))
    .init()
```
And I start it as `node app.js`

**Comment by zixia** [comment]
@ax4 

Had you tried install a NPM version of chrome-driver?

like run this:

```shell
npm install chrome-driver
```

**Comment by brianwfl88** [comment]
Hi @zixia,

I have noticed my console is trying to load chrome-driver.

However you suggest to install chromedriver. Will that be an issue? Since "chrome-driver" and "chromedriver" is a different thing and "chrome-driver" could not be found in npm repository.

**Comment by zixia** [comment]
It should be `chromedriver`. 

`chrome-driver` is my typo, sorry for that.

Did `npm install chromedriver` fix this issue for you?

**Comment by ax4** [comment]
Hi @zixia, 

my chromedriver was installed by `npm install -g chromedriver `, and it was added to PATH by it.

Although I tried `npm install chromedriver` this time, but it still won't work. 

Plus: my wechaty is also installed from npm, and it's on version 0.7.17 (described by the package itself)

**Comment by brianwfl88** [comment]
I have added "E:\wamp\www\GitHub\wechaty\node_modules\chromedriver\bin" into both System Path and User Path.

Some related issue with chrome driver on binary path: [https://github.com/angular/protractor/issues/950](url)


`INFO Wechaty v0.7.17 initializing...
WARN PuppetWebBrowserDriver valid() getSession() catch() rejected: unknown error
: cannot find Chrome binary
  (Driver info: chromedriver=2.27.440174 (e97a722caafc2d3a8b807ee115bfb307f7d2cf
d9),platform=Windows NT 6.1.7601 SP1 x86_64)
WARN PuppetWebBrowserDriver valid() driver.getSession() exception: unknown error
: cannot find Chrome binary
  (Driver info: chromedriver=2.27.440174 (e97a722caafc2d3a8b807ee115bfb307f7d2cf
d9),platform=Windows NT 6.1.7601 SP1 x86_64)
WARN PuppetWebBrowserDriver initChromeDriver() got invalid driver
WARN PuppetWebBrowserDriver initChromeDriver() with retry: 1
ERR Config ###########################
ERR Config unhandledRejection: WebDriverError: unknown error: cannot find Chrome
 binary
  (Driver info: chromedriver=2.27.440174 (e97a722caafc2d3a8b807ee115bfb307f7d2cf
d9),platform=Windows NT 6.1.7601 SP1 x86_64) [object Promise]
ERR Config ###########################
ERR Config unhandledRejection::catch(unknown error: cannot find Chrome binary
  (Driver info: chromedriver=2.27.440174 (e97a722caafc2d3a8b807ee115bfb307f7d2cf
d9),platform=Windows NT 6.1.7601 SP1 x86_64))
(node:6460) PromiseRejectionHandledWarning: Promise rejection was handled asynch
ronously (rejection id: 1)
WARN PuppetWebBrowserDriver valid() getSession() catch() rejected: unknown error
: cannot find Chrome binary
  (Driver info: chromedriver=2.27.440174 (e97a722caafc2d3a8b807ee115bfb307f7d2cf
d9),platform=Windows NT 6.1.7601 SP1 x86_64)
WARN PuppetWebBrowserDriver valid() driver.getSession() exception: unknown error
: cannot find Chrome binary
  (Driver info: chromedriver=2.27.440174 (e97a722caafc2d3a8b807ee115bfb307f7d2cf
d9),platform=Windows NT 6.1.7601 SP1 x86_64)
WARN PuppetWebBrowserDriver initChromeDriver() got invalid driver
WARN PuppetWebBrowserDriver initChromeDriver() with retry: 2
ERR Config ###########################
ERR Config unhandledRejection: WebDriverError: unknown error: cannot find Chrome
 binary
  (Driver info: chromedriver=2.27.440174 (e97a722caafc2d3a8b807ee115bfb307f7d2cf
d9),platform=Windows NT 6.1.7601 SP1 x86_64) [object Promise]
ERR Config ###########################
ERR Config unhandledRejection::catch(unknown error: cannot find Chrome binary
  (Driver info: chromedriver=2.27.440174 (e97a722caafc2d3a8b807ee115bfb307f7d2cf
d9),platform=Windows NT 6.1.7601 SP1 x86_64))
(node:6460) PromiseRejectionHandledWarning: Promise rejection was handled asynch
ronously (rejection id: 2)
WARN PuppetWebBrowserDriver valid() getSession() catch() rejected: unknown error
: cannot find Chrome binary
  (Driver info: chromedriver=2.27.440174 (e97a722caafc2d3a8b807ee115bfb307f7d2cf
d9),platform=Windows NT 6.1.7601 SP1 x86_64)
WARN PuppetWebBrowserDriver valid() driver.getSession() exception: unknown error
: cannot find Chrome binary
  (Driver info: chromedriver=2.27.440174 (e97a722caafc2d3a8b807ee115bfb307f7d2cf
d9),platform=Windows NT 6.1.7601 SP1 x86_64)
WARN PuppetWebBrowserDriver initChromeDriver() got invalid driver
WARN PuppetWebBrowserDriver initChromeDriver() with retry: 3
ERR Config ###########################
ERR Config unhandledRejection: WebDriverError: unknown error: cannot find Chrome
 binary
  (Driver info: chromedriver=2.27.440174 (e97a722caafc2d3a8b807ee115bfb307f7d2cf
d9),platform=Windows NT 6.1.7601 SP1 x86_64) [object Promise]
ERR Config ###########################
ERR Config unhandledRejection::catch(unknown error: cannot find Chrome binary
  (Driver info: chromedriver=2.27.440174 (e97a722caafc2d3a8b807ee115bfb307f7d2cf
d9),platform=Windows NT 6.1.7601 SP1 x86_64))
(node:6460) PromiseRejectionHandledWarning: Promise rejection was handled asynch
ronously (rejection id: 3)
WARN PuppetWebBrowserDriver valid() getSession() catch() rejected: unknown error
: cannot find Chrome binary
  (Driver info: chromedriver=2.27.440174 (e97a722caafc2d3a8b807ee115bfb307f7d2cf
d9),platform=Windows NT 6.1.7601 SP1 x86_64)
WARN PuppetWebBrowserDriver valid() driver.getSession() exception: unknown error
: cannot find Chrome binary
  (Driver info: chromedriver=2.27.440174 (e97a722caafc2d3a8b807ee115bfb307f7d2cf
d9),platform=Windows NT 6.1.7601 SP1 x86_64)
WARN PuppetWebBrowserDriver initChromeDriver() got invalid driver
WARN PuppetWebBrowserDriver initChromeDriver() not valid after retry: 4 times: E
rror: initChromeDriver() got invalid driver
    at BrowserDriver.<anonymous> (E:\wamp\www\GitHub\wechaty\node_modules\wechat
y\dist\src\puppet-web\browser-driver.js:116:35)
    at Generator.next (<anonymous>)
    at fulfilled (E:\wamp\www\GitHub\wechaty\node_modules\wechaty\dist\src\puppe
t-web\browser-driver.js:4:58)
    at process._tickCallback (internal/process/next_tick.js:103:7)
ERR PuppetWebBrowser init() exception: initChromeDriver() got invalid driver
ERR PuppetWebBrowser quit() this.driver.quit() exception unknown error: cannot f
ind Chrome binary
  (Driver info: chromedriver=2.27.440174 (e97a722caafc2d3a8b807ee115bfb307f7d2cf
d9),platform=Windows NT 6.1.7601 SP1 x86_64)
ERR Config ###########################
ERR Config unhandledRejection: WebDriverError: unknown error: cannot find Chrome
 binary
  (Driver info: chromedriver=2.27.440174 (e97a722caafc2d3a8b807ee115bfb307f7d2cf
d9),platform=Windows NT 6.1.7601 SP1 x86_64) [object Promise]
ERR Config ###########################
ERR Config unhandledRejection::catch(unknown error: cannot find Chrome binary
  (Driver info: chromedriver=2.27.440174 (e97a722caafc2d3a8b807ee115bfb307f7d2cf
d9),platform=Windows NT 6.1.7601 SP1 x86_64))
(node:6460) PromiseRejectionHandledWarning: Promise rejection was handled asynch
ronously (rejection id: 4)
ERR PuppetWeb initBrowser() exception: initChromeDriver() got invalid driver
ERR PuppetWeb init() exception: Error: initChromeDriver() got invalid driver
    at BrowserDriver.<anonymous> (E:\wamp\www\GitHub\wechaty\node_modules\wechat
y\dist\src\puppet-web\browser-driver.js:116:35)
    at Generator.next (<anonymous>)
    at fulfilled (E:\wamp\www\GitHub\wechaty\node_modules\wechaty\dist\src\puppe
t-web\browser-driver.js:4:58)
    at process._tickCallback (internal/process/next_tick.js:103:7)
ERR Wechaty init() exception: initChromeDriver() got invalid driver
ERR Config ###########################
ERR Config unhandledRejection: Error: initChromeDriver() got invalid driver [obj
ect Promise]
ERR Config ###########################
ERR Config unhandledRejection::catch(initChromeDriver() got invalid driver)
(node:6460) PromiseRejectionHandledWarning: Promise rejection was handled asynch
ronously (rejection id: 6)
WARN PuppetWebBrowserDriver valid() driver.getSession() timeout(halt?)
WARN PuppetWebBrowserDriver valid() driver.getSession() timeout(halt?)
WARN PuppetWebBrowserDriver valid() driver.getSession() timeout(halt?)
WARN PuppetWebBrowserDriver valid() driver.getSession() timeout(halt?)`

**Comment by zixia** [comment]
@brianwfl88 According to the error message: `unknown error : cannot find Chrome binary`

This time it seems that you need to install **chrome** browser.

**Comment by ax4** [comment]
@zixia Anyway, thank you so much for your help! I really appreciate it! 

Your docker solution is awesome! I'll keep going on bot development with docker solution, as well as figure out how it works on win32 (hopefully someday😅). 

BTW, I was so much impressed by your Wechaty 101 Video. Honestly, I have so much essentials to learn - from Docker thru CI, and Testing thru Building. Thanks to Wechaty project! It inspires me a lot!

**Comment by zixia** [comment]
@ax4 Thanks for loving Wechaty, appreciate it. :)

**Comment by brianwfl88** [comment]
hi @zixia ,

The program is running fine now on another windows machine, the chromedriver path must use **.exe** in the '**lib**' folder. 

After that, the demo script works fine. Thank you for your help.

**Comment by zixia** [comment]
@brianwfl88 Great to hear that, Cheers!
